File: src/main/java/engine/feature/primitives/Rectangle.java
package engine.feature.primitives;

import com.jogamp.opengl.GL4;
import com.jogamp.opengl.util.texture.Texture;
import engine.core.OpenGlObject;
import engine.util.color.ColorUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.awt.*;
import java.util.Arrays;


public class Rectangle extends OpenGlObject implements Primitive {

    public static float[] RECTANGLE_BUFFER = new float[]{
            0f, 1f,
            1f, 0f,
            0f, 0f,
            0f, 1f,
            1f, 1f,
            1f, 0f};

    public static float[] RECTANGLE_REVERSED_BUFFER = new float[]{
            0f, 0f,
            1f, 1f,
            0f, 1f,
            0f, 0f,
            1f, 0f,
            1f, 1f};

    private static int RECTANGLE_BUFFER_PARAMS_COUNT = 2;
    private static int RECTANGLE_VERTICES_COUNT = 6;

    private static String ERR_NOT_VALID_ATTRIB_BUFFER = "The buffer supplied is not valid for rectangle";

    public Rectangle(@NotNull final GL4 gl, final float width, final float height, final int textureId) {
        super(RECTANGLE_BUFFER_PARAMS_COUNT, RECTANGLE_VERTICES_COUNT, gl, new Dimension((int) width, (int) height), textureId);
    }

    public Rectangle(@NotNull final GL4 gl, final float posX, final float posY, final float width, final float height, final int textureId) {
        super(RECTANGLE_BUFFER_PARAMS_COUNT, RECTANGLE_VERTICES_COUNT, gl, posX, posY, new Dimension((int) width, (int) height), textureId);
    }

    public void init(@NotNull final String[] textureFilePaths,
                     final boolean texArray,
                     @NotNull final float[] attribDataArray) {

        if (isBufferValidForPrimitive(attribDataArray)) {
            super.initRenderData(textureFilePaths, texArray, RECTANGLE_BUFFER, attribDataArray);
        } else {
            throw new IllegalArgumentException(ERR_NOT_VALID_ATTRIB_BUFFER);
        }
    }

    public void init(@Nullable final Texture texture,
                     @NotNull final float[] attribDataArray) {

        if (isBufferValidForPrimitive(attribDataArray)) {
            super.initRenderData(texture, RECTANGLE_BUFFER, attribDataArray);
        } else {
            throw new IllegalArgumentException(ERR_NOT_VALID_ATTRIB_BUFFER);
        }
    }

    public void init(@NotNull final Color color) {
        super.initRenderData(null, RECTANGLE_BUFFER, ColorUtil.getBufferForColor(RECTANGLE_VERTICES_COUNT, color));
    }

    @Override
    public void initRenderData(@NotNull final String[] textureFilePaths, final boolean texArray, @NotNull final float[]... dataArrays) {
        validateSuppliedData(dataArrays);
        super.initRenderData(textureFilePaths, texArray, dataArrays);
    }

    @Override
    public void initRenderData(final Texture texture, @NotNull final float[]... dataArrays) {
        validateSuppliedData(dataArrays);
        super.initRenderData(texture, dataArrays);
    }

    @Override
    public void validateSuppliedData(@NotNull float[]... dataArrays) {

        final String ERR_NOT_VALID_BUFFER_COUNT = "Rectangle primitive can only have 2 buffers (vertex and attrib).";
        if (dataArrays.length != RECTANGLE_BUFFER_PARAMS_COUNT)
            throw new IllegalArgumentException(ERR_NOT_VALID_BUFFER_COUNT);

        final String ERR_NOT_VALID_VERTEX_BUFFER = "Rectangular vertex data should be supplied as the first argument for rectangle primitive. \n" +
                "You can use Rectangle.init() instead";
        if (!Arrays.equals(RECTANGLE_BUFFER, dataArrays[0]))
            throw new IllegalArgumentException(ERR_NOT_VALID_VERTEX_BUFFER);

        if (!isBufferValidForPrimitive(dataArrays[1]))
            throw new IllegalArgumentException(ERR_NOT_VALID_ATTRIB_BUFFER);
    }

    @Override
    public boolean isBufferValidForPrimitive(@NotNull float[] dataArray) {
        return dataArray.length % RECTANGLE_VERTICES_COUNT == 0;
    }
}

File: src/main/java/engine/feature/primitives/Rectangle.kt
package engine.feature.primitives

import com.jogamp.opengl.GL4
import com.jogamp.opengl.util.texture.Texture
import engine.core.OpenGlObject
import engine.util.color.ColorUtil

import java.awt.*
import java.util.Arrays


class Rectangle : OpenGlObject, Primitive {

    constructor(gl: GL4, width: Float, height: Float, textureId: Int) :
            super(RECTANGLE_BUFFER_PARAMS_COUNT, RECTANGLE_VERTICES_COUNT, gl, Dimension(width.toInt(), height.toInt()), textureId)

    constructor(gl: GL4, posX: Float, posY: Float, width: Float, height: Float, textureId: Int) :
            super(RECTANGLE_BUFFER_PARAMS_COUNT, RECTANGLE_VERTICES_COUNT, gl, posX, posY, Dimension(width.toInt(), height.toInt()), textureId)

    fun init(textureFilePaths: Array<String>, texArray: Boolean, attribDataArray: FloatArray) {
        if (isBufferValidForPrimitive(attribDataArray)) {
            super.initRenderData(textureFilePaths, texArray, RECTANGLE_BUFFER, attribDataArray)
        } else {
            throw IllegalArgumentException(ERR_NOT_VALID_ATTRIB_BUFFER_MSG)
        }
    }

    fun init(texture: Texture?, attribDataArray: FloatArray) {
        if (isBufferValidForPrimitive(attribDataArray)) {
            super.initRenderData(texture, RECTANGLE_BUFFER, attribDataArray)
        } else {
            throw IllegalArgumentException(ERR_NOT_VALID_ATTRIB_BUFFER_MSG)
        }
    }

    fun init(color: Color) {
        super.initRenderData(null, RECTANGLE_BUFFER, ColorUtil.getBufferForColor(RECTANGLE_VERTICES_COUNT, color))
    }

    override fun initRenderData(textureFilePaths: Array<String>, texArray: Boolean, vararg dataArrays: FloatArray) {
        validateSuppliedData(*dataArrays)
        super.initRenderData(textureFilePaths, texArray, *dataArrays)
    }

    override fun initRenderData(texture: Texture?, vararg dataArrays: FloatArray) {
        validateSuppliedData(*dataArrays)
        super.initRenderData(texture, *dataArrays)
    }

    override fun validateSuppliedData(vararg dataArrays: FloatArray) {
        var errMsg = "Rectangle primitive can only have 2 buffers (vertex and attrib)."
        if (dataArrays.size != RECTANGLE_BUFFER_PARAMS_COUNT)
            throw IllegalArgumentException(errMsg)

        errMsg = "Rectangular vertex data should be supplied as the first argument for rectangle primitive. \n" + "You can use Rectangle.init() instead"
        if (!Arrays.equals(RECTANGLE_BUFFER, dataArrays[0]))
            throw IllegalArgumentException(errMsg)

        if (!isBufferValidForPrimitive(dataArrays[1]))
            throw IllegalArgumentException(ERR_NOT_VALID_ATTRIB_BUFFER_MSG)
    }

    override fun isBufferValidForPrimitive(dataArray: FloatArray): Boolean {
        return dataArray.size % RECTANGLE_VERTICES_COUNT == 0
    }

    companion object {
        var RECTANGLE_BUFFER = floatArrayOf(0f, 1f, 1f, 0f, 0f, 0f, 0f, 1f, 1f, 1f, 1f, 0f)
        var RECTANGLE_REVERSED_BUFFER = floatArrayOf(0f, 0f, 1f, 1f, 0f, 1f, 0f, 0f, 1f, 0f, 1f, 1f)

        private const val RECTANGLE_BUFFER_PARAMS_COUNT = 2
        private const val RECTANGLE_VERTICES_COUNT = 6

        private const val ERR_NOT_VALID_ATTRIB_BUFFER_MSG = "The buffer supplied is not valid for rectangle"
    }
}
